File: breaking_sticks.py
#!/bin/python

import sys
from math import sqrt
MAX = 1000001


def prime2(n):
    

def sieve_of_eratosthenes():
    global sieve
    sieve = [1]*MAX
    sieve[0] = 0
    sieve[1] = 0
    lim = int(sqrt(MAX))
    for i in range(2,lim):
        if(sieve[i]):
            for j in range(i*i,MAX,i):
                sieve[j] = 0
    global primes
    primes = [2]
    for i in range(3,MAX):
        if(sieve[i]):
            primes += [i]

def prime(n):
    if(n >= MAX):
        for i in range(len(primes)):
            if(n % primes[i] == 0):
                return False
    else:
        return sieve[n]
    return False


def getDivisions(n):

    if(n == 1):
        return 1

    temp = n 
    if(temp % 2 == 0):
        sumVal = 0

        while(n % 2 == 0):
            sumVal += n
            n /= 2
        
        ans = 0
        sumVal += n 

        while(not prime(n) and n > 1):
            lim = int(sqrt(n))

            for i in range(3,lim+1,2):
                if(n % i == 0):
                    ans = n/i
                    break

            sumVal += ans
            n = ans
        sumVal += 1
        return sumVal
    else:

        sumVal = 0
        n = temp
        sumVal += n
        ans = 0

        while(not prime(n) and n > 1):
            lim = int(sqrt(n))

            for i in range(3,lim+1,2):
                if(n % i == 0):
                    ans = n/i
                    break

            sumVal += ans
            # print "First Odd divisor",ans
            n = ans
        sumVal += 1
        return sumVal

def longestSequence(a):
    #  Return the length of the longest possible sequence of moves.
    sumVal = 0
    for i in a:
        ans = getDivisions(i)
        sumVal += ans
    return sumVal


if __name__ == "__main__":
    n = int(raw_input().strip())
    a = map(long, raw_input().strip().split(' '))
    sieve_of_eratosthenes()
    result = longestSequence(a)
    print result

File: World-Codesprint12/breaking_sticks.cpp
#include <bits/stdc++.h>

using namespace std;

int prime(unsigned long long int n)
{
    long lim = sqrt(n)+1;
    for(int i = 3; i <= lim; i++)
    {
        if(n % i == 0)
        {
            return i;
        }
    }
    return 0;
}

unsigned long long int getDivisions(unsigned long long int n)
{
    // cout<<n<<"\n Called";
    if(n == 1)
        return 1;

    unsigned long long int temp = n;
    if(temp % 2 == 0)
    {
        unsigned long long int sum = 0;
        while(n % 2 == 0)
        {
            sum += n;
            n /= 2;
        }
        int ans = 0;
        // cout<<"\nIntermed n "<<n<<endl;
        if(n!=1)
            sum += n;
        while(prime(n) > 0 && n > 1)
        {
            ans = prime(n);
            sum += n/ans;
            n = n/ans;
            // cout<<"\nFirst Divisor of oddd: "<<ans<<endl;
        }
        sum += 1;
        return sum;
    }
    else if(prime(n) == 0)
    {
        return n+1;
    }
    else
    {
        unsigned long long int sum = 0;
        n = temp;
        sum += n;
        // cout<<"CALLING : "<<temp<<endl;
        int ans = 0;
        
        while(prime(n) > 0 && n > 1)
        {
            // cout<<"\nPRIME CHECKING: "<<n<<endl;
            ans = prime(n);
            sum += n/ans;
            n = n/ans;
            // cout<<"\nFirst Divisor of oddd: "<<ans<<endl;
        }
        sum += 1;
        return sum;
    }
}

unsigned long long int longestSequence(vector <unsigned long long int> a) {
    //  Return the length of the longest possible sequence of moves.
    unsigned long long int sum = 0;
    for(int i = 0; i < a.size() ; i++)
    {
        unsigned long long int n = a[i];
        unsigned long long int ans = getDivisions(n);
        // cout<<ans<<endl;
        sum += ans;
    }
    return sum;
}

int main() {
    int n;
    cin >> n;
    vector<unsigned long long int> a(n);
    for(int a_i = 0; a_i < n; a_i++){
       cin >> a[a_i];
    }
    // boolSieve = sieve_of_eratosthenes();
    unsigned long long int result = longestSequence(a);
    cout << result << endl;
    return 0;
}
